### ISSUE TITLE
[FIX] strapi-v4 - data provider - deep data normalization

### ISSUE THREAD

**Issue by lukassos** [pankod/refine#1485]
**Is your feature request related to a problem? Please describe.**
Populated data from strapi-v4 data  provider are not normalized, which makes hard to work with them via ts interfaces. 

**Describe the solution you'd like**
Use deep traversal like the one proposed here : https://forum.strapi.io/t/discussion-regarding-the-complex-response-structure-for-rest-graphql-developer-experience/13400/9
```js
const normalize = (data) => {
  const isObject = (data) =>
    Object.prototype.toString.call(data) === '[object Object]'
  const isArray = (data) =>
    Object.prototype.toString.call(data) === '[object Array]'

  const flatten = (data) => {
    if (!data.attributes) return data

    return {
      id: data.id,
      ...data.attributes
    }
  }

  if (isArray(data)) {
    return data.map((item) => normalize(item))
  }

  if (isObject(data)) {
    if (isArray(data.data)) {
      data = [...data.data]
    } else if (isObject(data.data)) {
      data = flatten({ ...data.data })
    } else if (data.data === null) {
      data = null
    } else {
      data = flatten(data)
    }

    for (const key in data) {
      data[key] = normalize(data[key])
    }

    return data
  }

  return data
}
```

**Describe alternatives you've considered**
Alternative would be to normalize data after you get query data. This might be too late if you load it as useTable which expect slightly modified data already.

**Additional context**
Read more at  : https://forum.strapi.io/t/discussion-regarding-the-complex-response-structure-for-rest-graphql-developer-experience/13400

**Comment by omeraplak** [comment]
Hey @lukassos ,
Looks beautiful! Thanks. Would you like to implement this for us?

**Comment by lukassos** [comment]
@omeraplak 
Already did some work and marked it as draft PR. may you have a look if it suits you ? 
Some tests will be necessary maybe dough.

**Comment by omeraplak** [comment]
Hey @lukassos ,
Great start. Congrats! I just saw that you haven't used the `normalizeData` function at all yet. I guess you haven't reached that stage yet.

**Comment by lukassos** [comment]
Hi @omeraplak 
Yeah I haven`t written any tests or usages if you are thinking about that, but this implementation works fine in our local environment after fetching data (this is sub optimal for lots of reasons). 

Implementing this in custom strapi api is also not a solution due to higher load on servers. 

So the usage should IMHO be in the data provider itself before we call on any other observers and start working with data. 
As you can see the interface is unchanged so usage at [line 123](https://github.com/pankod/refine/blob/1c1615fe2bf920a27cc1c120e2173952ca81fd5c/packages/strapi-v4/src/dataProvider.ts#L123) is exactly how I would like it to be used. :)

**Comment by omeraplak** [comment]
Hey @lukassos ,
First of all, many thanks.

1. We have documentation specific to Strapi v4. https://refine.dev/docs/guides-and-concepts/data-provider/strapi-v4/
Could you add the necessary explanations for this document?

2. We need to update the unit tests but I guess we can do that part. If you've used [nock](https://github.com/nock/nock) before (which is very easy) you can update the tests too.

**Comment by lukassos** [comment]
We should probably change the title to [FIX]?? 
This fixes general functionality of internal function in dataProvider nothing more. At the moment it does not work as expected from data normalization of strapi-v4 as only one level is normalized oll deeper  levels are ignored (does not traverse). So what you have in the PR is quick solution. 

------------

As for the points @omeraplak 
1. Writing documentation for dataProvider internal function is probably out of scope for guides and example. Not sure where to put it, maybe start new section or enhance the [Populate Realtions](https://refine.dev/docs/guides-and-concepts/data-provider/strapi-v4/#relations-population)? 

   If we export the function for users of refine-strapi-v4 then we can mention the function here. 

   What is missing in the docs is usage of multilevel relations, eg : 
   ```js
   const { tableProps } = useTable<IPost>({
    metaData: {
        populate: [ "category", "cover", "author", "author.user", "author.media", "author.media.avatar" ]
    },
   }); 
   ```

   IMHO good example would be to add multilevel field to post type, eg. author pointing to "plugin::users-permissions.user" table (by default included in every strapi-v4 instance) and some strapi component like media file. Then we can show exactly what does it mean to have multilevel data from strapi GET:findMany and refines normalized data structure it expects in TS implementation. 

2. Cannot promise any time-window for this :/ maybe is all I can provide here. Nock seems easy enough though.

**Comment by omeraplak** [comment]
Hey @lukassos ,
Hmm, I get it. Should we leave Refine's `strapi-v4` package as default? Because it can be a little difficult to change this default behavior and explain it to people.
If you want, publish it as a new package from your GitHub account and we'll mention to your Github repo/npm package in our documentation. How does this sound?

**Comment by IcebearInSvalbard** [comment]
I don't think this is something that needs a separate plugin. 
No sure I understand what you mean by _"leave strapi-v4 as default"_ ?? But from what I see you have internal function that does something to data schema which comes from strapi-4 (different from previous strapi schema) and it has no documentation you are doing this, which might be your problem here. My problem is that I get malformed data in mixed schema. 

AFAIK documentation can be as easy as one sentence or two example json outputs comparing (server output, data provider normalized output). 
 
What is really my dilemma here is that normalizeData flatens only getList and getMany data, but leaves all other in schema that can have arrays of objects represented as `{data[object]}` instead of `[object]` in nested populated fields. 

IMHO when normalization is applied it should be global effect but instead of nice simple json folowing one type of schema  
```json
{
  data:[
    {
      id:..
      content:...
      authors:[
        {...author object...}
      ]
    }
  ]
}
```
we get something mixed 

```json
{
  data:[
    {
      id:..
      content:...
      authors:{
       data: [
          {...author object...}
        ]
      }
    }
  ]
}
```

One would expect you would either normalize all the data or nothing. What do you think ? 

What is expected behavior by your standards ?
Because I get it that we might have different opinions on what is the normal expected schema here. 


And no I don't have no intentions to obfuscate plugins and npm packages when you are doing exactly the same thing. Goes against DRY principle. Maybe we can make another function that would traverse data only when triggered with some kind of flag but that goes against SOLID .. thus .. I'm in a pickle here and so are you as I can see? :D

**Comment by lukassos** [comment]
I don't think this is something that needs a separate plugin. 
Not sure I understand what you mean by _"leave strapi-v4 as default"_ ?? But from what I see you have internal function that does something to data schema which comes from strapi-4 (different from previous strapi schema) and it has no documentation you are doing this, which might be your problem here. My problem is that I get malformed data in mixed schema. 

AFAIK documentation can be as easy as one sentence or two example json outputs comparing (server output, data provider normalized output). 
 
What is really my dilemma here is that normalizeData flatens only getList and getMany data, but leaves all other in schema that can have arrays of objects represented as `{data[object]}` instead of `[object]` in nested populated fields. 

IMHO when normalization is applied it should be global effect but instead of nice simple json folowing one type of schema
```json
{
  "data":[
    {
      "id":..
      "content":...
      "authors":[
        {
          //...author object...
          "id":...
          ...authors props...
        }

      ]
    }
  ]
}
```
we get something mixed 

```json
{
  "data":[
    {
      "id":..
      "content":...
      "authors":{
       "data": [
          {
            //...author object...
            "id":...
            "attributes":{
                ...authors props...
            }
          }
        ]
      }
    }
  ]
}
```
So normalizeData flattens attributes and topmost data but leaves all nested related objects in strapi-v4 schema. 
One would expect you would either normalize all the data or nothing. What do you think ? 

What is expected behavior by your standards ?
Because I get it that we might have different opinions on what is the normal expected schema here. 


And no I don't have no intentions to obfuscate plugins and npm packages when you are doing exactly the same thing. Goes against DRY principle. Maybe we can make another function that would traverse data only when triggered with some kind of flag but that goes against SOLID .. thus .. I'm in a pickle here and so are you as I can see? :D

**Comment by omeraplak** [comment]
Hey @lukassos ,
I missed what we did to normalize. Thanks a lot for your warning. Now everything is clearer to me. Thanks so much for your passion!

What we have to do now is
- Fixing tests (I'll do that)
- To talk about this in our document.

You can create a new section for the document. I'll also try to set the "plugin::users-permissions.user" plugin today as you suggested. Is there anything I missed?

**Comment by omeraplak** [comment]
There was no situation that needed to be fixed in the tests. I updated our own `strapi-v4` example. Then we are waiting for them to add it to the document, then we will release it! thanks a lot

**Comment by salihozdemir** [comment]
Hey @lukassos 👋 ,

Thanks for your contribution. After this point, we will make a new release by making additions to what you have done.
When finished, we would appreciate it if you could review the changes and make sure they are correct.

**Comment by omeraplak** [comment]
Hey,
We've released `@pankod/refine-strapi-v4@3.1.7`

**Comment by lukassos** [comment]
Nice
I'll try to make time for this later today and let you know if it works the
way it should
Thank you
